NuGet fails because of invalid characters in User-Agent header

### NuGet Product Used

dotnet.exe, NuGet SDK

### Product Version

9.0.100-preview.4.24267.66

### Worked before?

No

### Impact

I'm unable to use this version

### Repro Steps & Context

1. Install NixOS latest version (24.11 Vicuña).
2. Attempt to restore nuget packages for a project.
3. Unable to restore packages due to special character in header since the release name (Vicuña) is added to the HTTP header.

### Verbose Logs

```shell
/nix/store/lax73wx0qp7v2wpydc0bii5hszj1dys5-dotnet-core-combined/sdk/9.0.100-preview.4.24267.66/MSBuild.dll -nologo -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/nix/store/jd1kkgy9psy5km4m6k0drbzd7i3vqga6-dotnet-sdk-9.0.100-preview.4.24267.66/sdk/9.0.100-preview.4.24267.66/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/nix/store/jd1kkgy9psy5km4m6k0drbzd7i3vqga6-dotnet-sdk-9.0.100-preview.4.24267.66/sdk/9.0.100-preview.4.24267.66/dotnet.dll -maxcpucount -target:Restore -tlp:default=auto -verbosity:m -verbosity:diag ./m3u8-dl.sln
    Determining projects to restore...
    /home/ggg/Git/GGG-KILLER/m3u8-dl/src/m3u8-dl.csproj : error NU1301: Unable to load the service index for source https://api.nuget.org/v3/index.json.
    Failed to restore /home/ggg/Git/GGG-KILLER/m3u8-dl/src/m3u8-dl.csproj (in 5.76 sec).
    /home/ggg/Git/GGG-KILLER/m3u8-dl/test/m3u8-dl.Test.csproj : error NU1301: Unable to load the service index for source https://api.nuget.org/v3/index.json.
    /home/ggg/Git/GGG-KILLER/m3u8-dl/src/m3u8-dl.csproj : error NU1301: Unable to load the service index for source https://api.nuget.org/v3/index.json.
    Failed to restore /home/ggg/Git/GGG-KILLER/m3u8-dl/test/m3u8-dl.Test.csproj (in 11.61 sec).
```

--- a/src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs
+++ b/src/NuGet.Core/NuGet.Build.Tasks/BuildTasksUtility.cs
@@ -5,9 +5,6 @@ using System;
 using System.Collections.Generic;
 using System.IO;
 using System.Linq;
-#if IS_CORECLR
-using System.Runtime.InteropServices;
-#endif
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.Build.Framework;
@@ -31,7 +28,6 @@ using NuGet.ProjectManagement;
 using NuGet.Shared;
 using static NuGet.Shared.XmlUtility;
 using System.Globalization;
-using System.Collections;
 #endif
 
 namespace NuGet.Build.Tasks
@@ -169,8 +165,7 @@ namespace NuGet.Build.Tasks
 
                 // Set user agent string used for network calls
 #if IS_CORECLR
-                UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet .NET Core MSBuild Task")
-                    .WithOSDescription(RuntimeInformation.OSDescription));
+                UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet .NET Core MSBuild Task"));
 #else
                 // OS description is set by default on Desktop
                 UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet Desktop MSBuild Task"));

--- a/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/XPlatUtility.cs
+++ b/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/XPlatUtility.cs
@@ -67,8 +67,7 @@ namespace NuGet.CommandLine.XPlat
         public static void SetUserAgent()
         {
 #if IS_CORECLR
-            UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet xplat")
-                .WithOSDescription(RuntimeInformation.OSDescription));
+            UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet xplat"));
 #else
             UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet xplat"));
 #endif

--- a/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/XPlatUtility.cs
+++ b/src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/XPlatUtility.cs
@@ -6,12 +6,6 @@
 using System.Globalization;
 using System;
 using System.IO;
-using System.Linq;
-
-#if IS_CORECLR
-using System.Runtime.InteropServices;
-#endif
-
 using NuGet.Common;
 using NuGet.Configuration;
 using NuGet.Protocol.Core.Types;
@@ -66,11 +60,7 @@ namespace NuGet.CommandLine.XPlat
 
         public static void SetUserAgent()
         {
-#if IS_CORECLR
             UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet xplat"));
-#else
-            UserAgent.SetUserAgentString(new UserAgentStringBuilder("NuGet xplat"));
-#endif
         }
 
         internal static ISettings ProcessConfigFile(string configFile)

--- a/src/NuGet.Core/NuGet.Protocol/UserAgentStringBuilder.cs
+++ b/src/NuGet.Core/NuGet.Protocol/UserAgentStringBuilder.cs
@@ -106,6 +106,10 @@ namespace NuGet.Protocol.Core.Types
                 return OSPlatform.OSX.ToString();
             }
 #if NETCOREAPP
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.FreeBSD))
+            {
+                return OSPlatform.FreeBSD.ToString();
+            }
             else if (OperatingSystem.IsBrowser())
             {
                 return "BROWSER";

--- a/src/NuGet.Core/NuGet.Protocol/UserAgentStringBuilder.cs
+++ b/src/NuGet.Core/NuGet.Protocol/UserAgentStringBuilder.cs
@@ -3,6 +3,7 @@
 
 using System;
 using System.Globalization;
+using System.Runtime.InteropServices;
 using NuGet.Packaging;
 
 namespace NuGet.Protocol.Core.Types
@@ -30,18 +31,15 @@ namespace NuGet.Protocol.Core.Types
 
             // Read the client version from the assembly metadata and normalize it.
             NuGetClientVersion = MinClientVersionUtility.GetNuGetClientVersion().ToNormalizedString();
+
+            _osInfo = GetOS();
         }
 
         public string NuGetClientVersion { get; }
 
+        [Obsolete("This value is now ignored")]
         public UserAgentStringBuilder WithOSDescription(string osInfo)
         {
-#if NETCOREAPP2_0_OR_GREATER
-            _osInfo = osInfo.Replace("(", @"\(", StringComparison.Ordinal).Replace(")", @"\)", StringComparison.Ordinal);
-#else
-            _osInfo = osInfo.Replace("(", @"\(").Replace(")", @"\)");
-#endif
-
             return this;
         }
 
@@ -53,8 +51,6 @@ namespace NuGet.Protocol.Core.Types
 
         public string Build()
         {
-            var osDescription = _osInfo ?? GetOSVersion();
-
             var clientInfo = _clientName;
             if (NuGetTestMode.Enabled)
             {
@@ -65,7 +61,7 @@ namespace NuGet.Protocol.Core.Types
                 clientInfo = DefaultNuGetClientName;
             }
 
-            if (string.IsNullOrEmpty(osDescription))
+            if (string.IsNullOrEmpty(_osInfo))
             {
                 return string.Format(
                     CultureInfo.InvariantCulture,
@@ -81,7 +77,7 @@ namespace NuGet.Protocol.Core.Types
                     UserAgentWithOSDescriptionTemplate,
                     clientInfo,
                     NuGetClientVersion,
-                    osDescription);
+                    _osInfo);
             }
             else
             {
@@ -90,26 +86,35 @@ namespace NuGet.Protocol.Core.Types
                     UserAgentWithOSDescriptionAndVisualStudioSKUTemplate,
                     _clientName,
                     NuGetClientVersion, /* NuGet version */
-                    osDescription, /* OS version */
+                    _osInfo, /* OS version */
                     _vsInfo);  /* VS SKU + version */
             }
         }
 
-        private string GetOSVersion()
+        internal static string GetOS()
         {
-            if (_osInfo == null)
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
             {
-#if !IS_CORECLR
-                // When not on CoreClr and no OSDescription was provided,
-                // we will set it ourselves.
-                _osInfo = Environment.OSVersion.ToString();
-#else
-                // When on CoreClr, one should use the .WithOSDescription() method to set it.
-                _osInfo = string.Empty;
-#endif
+                return OSPlatform.Windows.ToString();
             }
-
-            return _osInfo;
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                return OSPlatform.Linux.ToString();
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                return OSPlatform.OSX.ToString();
+            }
+#if NETCOREAPP
+            else if (OperatingSystem.IsBrowser())
+            {
+                return "BROWSER";
+            }
+#endif
+            else
+            {
+                return "UnknownOS";
+            }
         }
     }
 }

--- a/test/NuGet.Core.Tests/NuGet.Protocol.Tests/UserAgentStringBuilderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Protocol.Tests/UserAgentStringBuilderTests.cs
@@ -1,7 +1,8 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-using System.Net.Http;
+using FluentAssertions;
+using NuGet.Packaging;
 using NuGet.Protocol.Core.Types;
 using Xunit;
 using Xunit.Abstractions;
@@ -57,15 +58,19 @@ namespace NuGet.Protocol.Tests
         }
 
         [Fact]
-        public void UsesProvidedOSDescription()
+        public void AddsOSInfo()
         {
-            var osDescription = "OSName/OSVersion";
-            var builder = new UserAgentStringBuilder();
-            var userAgentString = builder.WithOSDescription(osDescription).Build();
-            var userAgentString2 = builder.WithOSDescription(osDescription).WithVisualStudioSKU("VS SKU/Version").Build();
+            var clientName = "clientName";
+            var clientVersion = MinClientVersionUtility.GetNuGetClientVersion().ToNormalizedString();
+            var os = UserAgentStringBuilder.GetOS();
+            var vs = "VS SKU/Version";
 
-            Assert.Contains($"({osDescription})", userAgentString);
-            Assert.Contains($"({osDescription}, VS SKU/Version)", userAgentString2);
+            var builder = new UserAgentStringBuilder(clientName);
+            var userAgentString = builder.Build();
+            var userAgentString2 = builder.WithVisualStudioSKU(vs).Build();
+
+            userAgentString.Should().Be($"{clientName}/{clientVersion} ({os})");
+            userAgentString2.Should().Be($"{clientName}/{clientVersion} ({os}, {vs})");
         }
 
         [Fact]
@@ -73,10 +78,10 @@ namespace NuGet.Protocol.Tests
         {
             var vsInfo = "VS SKU/Version";
             var builder = new UserAgentStringBuilder();
-            var userAgentString = builder.WithOSDescription("OSName/OSVersion").WithVisualStudioSKU(vsInfo).Build();
+            var userAgentString = builder.WithVisualStudioSKU(vsInfo).Build();
 
             _output.WriteLine(userAgentString);
-            Assert.Contains($"(OSName/OSVersion, {vsInfo})", userAgentString);
+            Assert.Contains($", {vsInfo})", userAgentString);
         }
 
         [Fact]
@@ -84,34 +89,14 @@ namespace NuGet.Protocol.Tests
         {
             var builder = new UserAgentStringBuilder();
 
-            var userAgentString = builder.WithOSDescription("OSName/OSVersion").WithVisualStudioSKU("VS SKU/Version").Build();
-            var userAgentString2 = builder.WithOSDescription("OSName/OSVersion").Build();
-            var userAgentString3 = builder.WithVisualStudioSKU("VS SKU/Version").Build();
-            var userAgentString4 = builder.Build();
+            var userAgentString = builder.WithVisualStudioSKU("VS SKU/Version").Build();
+            var userAgentString2 = builder.Build();
 
             _output.WriteLine("NuGet client version: " + builder.NuGetClientVersion);
             Assert.NotNull(builder.NuGetClientVersion);
             Assert.NotEmpty(builder.NuGetClientVersion);
             Assert.Contains(builder.NuGetClientVersion, userAgentString);
             Assert.Contains(builder.NuGetClientVersion, userAgentString2);
-            Assert.Contains(builder.NuGetClientVersion, userAgentString3);
-            Assert.Contains(builder.NuGetClientVersion, userAgentString4);
-        }
-
-        [Theory]
-        [InlineData("Custom Kernel (123")]
-        [InlineData("Custom Kernel 123)")]
-        public void Build_OsDescriptionWithUnmatchedParenthesis_IsValid(string osDescription)
-        {
-            // Arrange
-            UserAgentStringBuilder target = new();
-
-            // Act
-            string result = target.WithOSDescription(osDescription).Build();
-
-            // Assert
-            HttpRequestMessage httpRequest = new();
-            httpRequest.Headers.Add("User-Agent", result);
         }
     }
 }

